### PR TITLE
Opt-in for generation userid, requiring email addresses

### PR DESCRIPTION
**Pull request by blizzz**
This PR adds following two switches for creating new users:

* requiring email address
* auto-generation of userid

The motivation is to allow untechnical (sub)admins to allow to create users on an external backend. The user id does not play a role, but the email has to be enforced.

Therefore, there are no graphical switches. Bu the flags can be set via occ and they also do work with the local backend. For testing:

```
occ config:app:set settings newUser.requireEmail --value=yes
```

This flag sets the email field to required when creating a new user. The provsioning API errors with 110 if it is not provided.

```
occ config:app:set settings newUser.generateUserID --value=yes
```

This disables the new username field on the users page. The user will be appended to the list after creation with a random 10-character long ascii string. In case an unexisting userid could not be generated, the provisioning API will error with 111 after 10 attempts.

### CODE REVIEW

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 1]
```diff
@@ -223,6 +237,10 @@ public function addUser(string $userid,
 		$isAdmin = $this->groupManager->isAdmin($user->getUID());
 		$subAdminManager = $this->groupManager->getSubAdmin();
 
+		if(empty($userid) && (bool)$this->config->getAppValue('settings', 'newUser.generateUserID', false)) {
```
Comment (skjnldsv) [inline apps/provisioning_api/lib/Controller/UsersController.php]: Casting? Shouldn't we rather check against the exact value? '1'?
Or is it not a bad practice :)

**Review by skjnldsv — APPROVED** [review]
Code and tests are good. A small nitpick that is maybe not an issue! :)

File: apps/provisioning_api/lib/Controller/UsersController.php (hunk 1 above)
Comment (blizzz) [reply apps/provisioning_api/lib/Controller/UsersController.php]: Eh … yeah, whyever I was doing it this way. I guess I was carried away by my urge to have boolean meaning be treated boolean. Even the cast useless here. I'll change it :)

Comment (skjnldsv) [reply apps/provisioning_api/lib/Controller/UsersController.php]: @blizzz well, you're the php expert here ;)
So if you say it's fine, it's fine for me! :rocket:

**Comment by blizzz** [comment]
May I have a second reviewer, please?

**Review by skjnldsv — APPROVED** [review]
Do I count? :smile:

**Comment by wiswedel** [comment]
@juliushaertl @georgehrke @schiessle @vincchan 
Can someone please review this PR? We have to deliver.
Thx

**Comment by blizzz** [comment]
Meh, conflicts … who dared to merge other stuff first 😝 OK, gonna rebase.

> Do I count? smile

If this is one of your other personalities speaking, then, I think, yes, why not? 🙃 :clown_face:

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 2]
```diff
@@ -223,6 +237,10 @@ public function addUser(string $userid,
 		$isAdmin = $this->groupManager->isAdmin($user->getUID());
 		$subAdminManager = $this->groupManager->getSubAdmin();
 
+		if(empty($userid) && $this->config->getAppValue('settings', 'newUser.generateUserID', '0') === '1') {
```
Comment (georgehrke) [inline apps/provisioning_api/lib/Controller/UsersController.php]: Nitpick: we usually use 'yes' and 'no' (see the `enabled` configkey for example)

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 3]
```diff
@@ -275,6 +293,10 @@ public function addUser(string $userid,
 			$generatePasswordResetToken = true;
 		}
 
+		if ($email === '' && $this->config->getAppValue('settings', 'newUser.requireEmail', '0') === '1') {
```
Comment (georgehrke) [inline apps/provisioning_api/lib/Controller/UsersController.php]: see above

**Review by georgehrke — APPROVED** [review]
Tiny nitpick, otherwise good to go 👍

File: apps/provisioning_api/lib/Controller/UsersController.php (hunk 2 above)
Comment (georgehrke) [reply apps/provisioning_api/lib/Controller/UsersController.php]: we also don't use `getAppValue('settings'` anywhere, it's usually `getAppValue('core'`

Comment (blizzz) [reply apps/provisioning_api/lib/Controller/UsersController.php]: Indeed! Ok, will adjust, thank you!

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 4]
```diff
@@ -315,7 +337,7 @@ public function addUser(string $userid,
 				}
 			}
 
-			return new DataResponse();
+			return new DataResponse(['UserID' => $userid]);
```
Comment (nickvergessen) [inline apps/provisioning_api/lib/Controller/UsersController.php]: should we use `id`, so the key is the same as from `getUserData`?

Comment (skjnldsv) [reply apps/provisioning_api/lib/Controller/UsersController.php]: AH yes maybe we should!
@blizzz ?

Comment (blizzz) [reply apps/provisioning_api/lib/Controller/UsersController.php]: makes sense

Comment (blizzz) [reply apps/provisioning_api/lib/Controller/UsersController.php]: https://github.com/nextcloud/server/pull/16036

**Pull request merged**